### PR TITLE
Decouple functionality via events/signals

### PR DESCRIPTION
**Pull request by feliam**
WIP
* Executor.py refactor
* State search policy decoupled to its own class. Policy() So far only Random() is implemented
* Executor fully equipped with events. Main responsibilities: 
     - backup/restore states from _storage_
     - assign id numbers to new states and testcases
     - report all kid of events via signals
* abstractcpu now signals decode and execute as separated events
* Manticore object now needs to hook on some of the new events to generate testcases and others
* mappings.py moved to utils were it belongs. Yes mappings you belong there.
* events.py now handles forwarding/disable/enable of signals

Things to watch for:
 - Exceptions where are defined, where raised etc
 - Signals, propagation and re-entrance? add/del callbacks in callbacks?
 - Concurrence State.context vs Executor.context
 - If you don't know why is that there it needs doc or deletion

### CODE REVIEW

File: manticore/core/cpu/x86.py [hunk 1]
```diff
@@ -5620,6 +5633,102 @@ def VZEROUPPER(cpu):
 ################################################################################
 #Calling conventions
 
+class ABI:
```
Comment (mossberg) [inline manticore/core/cpu/x86.py]: it looks like the ABI abstraction refactor was undone here, and things are being added back?

File: manticore/core/cpu/abstractcpu.py [hunk 2]
```diff
@@ -293,13 +352,15 @@ def syscall_number(self):
         '''
         raise NotImplementedError
 
+from ...utils.event import Signal
+
 ############################################################################
-# Abstract cpu encapsulating common cpu methods used by platforms and executor.
+# Abstract cpu encapsulating common cpu methods used by models and executor.
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: seem "platforms" have been replaced by "models" in some places they shouldn't?

File: manticore/core/cpu/abstractcpu.py [hunk 3]
```diff
@@ -334,7 +409,7 @@ def __getstate__(self):
     def __setstate__(self, state):
         Cpu.__init__(self, state['regfile'], state['memory'])
         self._icount = state['icount']
-        return
+        return
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: nit, space

File: manticore/core/cpu/abstractcpu.py [hunk 4]
```diff
@@ -324,6 +385,20 @@ def __init__(self, regfile, memory):
...
+        #####################################################
+        # Signals
+        # signal handlers must have this signature:
+        # handler(cpu, *args, **kwargs)
+        self.will_decode_instruction = Signal()
+        self.will_execute_instruction = Signal()
+        self.did_execute_instruction = Signal()
+        self.will_emulate_instruction = Signal()
+        self.did_emulate_instruction = Signal()
+        self.will_read_register = Signal()
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: it seems like some of these are unused? can have documented use cases before adding these in?

File: manticore/core/cpu/abstractcpu.py [hunk 5]
```diff
@@ -479,30 +559,21 @@ def read_bytes(self, where, size):
...
     #######################################
     # Decoder
-    @abstractmethod
     def _wrap_operands(self, operands):
         '''
         Private method to decorate a capstone Operand to our needs. See Operand
         class
         '''
-        pass
+        raise NotImplemented
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: should be NotImplementedError

File: manticore/core/cpu/abstractcpu.py [hunk 6]
```diff
@@ -479,30 +559,21 @@ def read_bytes(self, where, size):
...
 
     def decode_instruction(self, pc):
         '''
-        This will decode an instruction from memory pointed by `pc` and store
-        it in self.instruction.
+        This will decode an instruction from memory pointed by @pc
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: was this intentional? i think backticks is the reST syntax?

File: manticore/core/cpu/abstractcpu.py [hunk 7]
```diff
@@ -518,7 +589,11 @@ def decode_instruction(self, pc):
                     if isinstance(c, Constant):
                         c = chr(c.value)
                     else:
+                        logger.info('Trying to execute instructions from invalid memory')
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: doesn't the existing logger call here already convey basically this information?

File: manticore/core/executor.py [hunk 8]
```diff
@@ -444,401 +382,172 @@ def generate_testcase(self, state, message = 'Testcase generated'):
...
-                    if current_state:
-                        assert solver.check(current_state.constraints)
-
-                    #execute until exception or finish or max consecutive instructions without exception 
-                    max_iters = self.dump_every + count
-
-                    # allow us to terminate manticore processes
-                    while not self.isShutdown():
-                        # Make sure current instruction is decoded so that hooks can access it
-                        current_state.cpu.decode_instruction(current_state.cpu.PC)
```
Comment (mossberg) [inline manticore/core/executor.py]: does this undo https://github.com/trailofbits/manticore/pull/205? should it?

File: manticore/core/cpu/abstractcpu.py [hunk 9]
```diff
@@ -570,27 +646,46 @@ def fallback_to_emulate(*operands):
...
+        self.instruction = instruction
         implementation(*instruction.operands)
+        self.instruction = None
         self._icount+=1
 
+        #broadcast event
+        self.did_execute_instruction(instruction)
+
+
+    def get_syscall_description(self):
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: this was mistakenly added back?

File: manticore/core/executor.py [hunk 10]
```diff
@@ -69,184 +46,276 @@ def newFunction(self, *args, **kw):
...
-        assert policy in ['random','adhoc','uncovered','dicount','icount','syscount','depth','bucket']
-        self.policy = policy
-
-        self._states = manager.dict()
-        self._test_number = manager.Value('i', 0 )
+        # Signals / Callbacks handlers will be invoked potentially at differeent 
+        # worker processes. State provides a local context to save data.
+
+        #Executor
+        self.will_start_run = Signal()
```
Comment (mossberg) [inline manticore/core/executor.py]: are we going a little signal crazy here? a lot of these seem unused. again, i think we should have documented use cases first

File: manticore/core/executor.py [hunk 11]
```diff
@@ -69,184 +46,276 @@ def newFunction(self, *args, **kw):
...
+            It needs a lock. Its used like this:
+
+            with executor.context() as context:
+                vsited = context['visited']
+                visited.append(state.cpu.PC)
+                context['visited'] = visited
+        '''
+        with self._lock:
+            yield self._shared_context
+        #FIXME What if we do a:
```
Comment (mossberg) [inline manticore/core/executor.py]: @feliam when you have this kind of fixme thing where you're unsure, can you bring it up explicitly? maybe mention it in the PR description? or in a comment? or slack message?

File: manticore/manticore.py [hunk 12]
```diff
@@ -16,25 +16,47 @@
...
+            elif func_name == 'backup':
+                self.saving_time += func_time
+            elif func_file.endswith('solver.py') and 'setstate' not in func_name and 'getstate' not in func_name and 'ckl' not in func_name:
+                self.solver_time += func_time
+
+
+
 def makeDecree(args):
     constraints = ConstraintSet()
-    platform = decree.SDecree(constraints, ','.join(args.programs))
```
Comment (mossberg) [inline manticore/manticore.py]: mistaken renames?

File: manticore/core/executor.py (hunk 8 above)
Comment (feliam) [inline manticore/core/executor.py]: will_execute_instruction signal now is called only when the instruction was successfully decoded. 
So _I think_ we don't need to force decoding here.

File: manticore/core/executor.py (hunk 10 above)
Comment (feliam) [inline manticore/core/executor.py]: I'm thinking maybe a gui that wants to report what's happening. But yeah! Most Signals around are     a request for comment basically. ( There are several workers that may start or not ). This is definitively a candidate for deletion.

File: manticore/utils/emulate.py [hunk 13]
```diff
@@ -218,11 +218,12 @@ def _step(self, instruction):
             val = self._cpu.read_register(reg)
             if issymbolic(val):
                 from ..core.cpu.abstractcpu import ConcretizeRegister
-                raise ConcretizeRegister(reg, "Concretizing for emulation.",
+                raise ConcretizeRegister(self._cpu, reg, "Concretizing for emulation.",
                                          policy='ONE') 
             self._emu.reg_write(self._to_unicorn_id(reg), val)
 
         # Bring in the instruction itself
+        instruction = self._cpu.decode_instruction(self._cpu.PC)
```
Comment (mossberg) [inline manticore/utils/emulate.py]: also `emulate_next` in test_unicorn.py should be updated

File: manticore/manticore.py [hunk 14]
```diff
@@ -505,7 +537,189 @@ def load_assertions(self, path):
...
                 self._assertions[pc] = ' '.join(line.split(' ')[1:])
 
+    def _backup_state_callback(self, state, state_id):
+        logger.debug("Backup state %r", state_id)
+
+    def _restore_state_callback(self, state, state_id):
+        logger.debug("Restore state %r", state_id)
+
+    def _terminate_state_callback(self, state, state_id, ex):
+        print "Terminate state", state.context
```
Comment (mossberg) [inline manticore/manticore.py]: rm debug print?

File: manticore/models/system.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+class OSException(Exception):
```
Comment (mossberg) [inline manticore/models/system.py]: what is this file for? models/ is for function models, so should it be somewhere else?

File: manticore/platforms/decree.py [hunk 16]
```diff
@@ -15,13 +15,15 @@
 import logging
 import random
 
-logger = logging.getLogger("PLATFORM")
+logger = logging.getLogger("MODEL")
```
Comment (mossberg) [inline manticore/platforms/decree.py]: can we fix all these model/platform naming issues?

File: manticore/manticore.py [hunk 17]
```diff
@@ -533,32 +747,92 @@ def run(self, procs=1, timeout=0):
...
+        self._executor.will_decode_instruction += self._decode_instruction_callback
+        self._executor.will_backup_state += self._backup_state_callback
+        self._executor.will_restore_state += self._restore_state_callback
+        self._executor.will_fork_state += self._fork_state_callback
+        self._executor.will_terminate_state += self._terminate_state_callback
+        self._executor.will_generate_testcase += self._generate_testcase_callback
+
         if self._hooks:
-            self._executor.will_execute_pc += self._hook_callback
+            self._executor.will_execute_state += self._hook_callback
```
Comment (mossberg) [inline manticore/manticore.py]: executor.will_execute_state is never declared so this breaks hooking

File: manticore/core/cpu/abstractcpu.py [hunk 18]
```diff
@@ -479,89 +554,98 @@ def read_bytes(self, where, size):
...
+        if not self.memory.access_ok(self.PC,'x'):
+            raise InvalidMemoryAccess(self.PC, 'x')
+
+        #broadcast event
+        self.will_decode_instruction()
+
+        instruction = self.decode_instruction(self.PC)
+
+        #broadcast event
+        self.will_execute_instruction(instruction)
```
Comment (mossberg) [inline manticore/core/cpu/abstractcpu.py]: is this the signal that hooks will be fired using? we need to set self.instruction before this so they can access the current instruction?

File: manticore/manticore.py [hunk 19]
```diff
@@ -505,7 +537,189 @@ def load_assertions(self, path):
...
+        state_visited = state.context.get('visited', set())
+        state_instructions_count = state.context.get('instructions_count', 0)
+        with self._executor.locked_context() as context:
+            executor_visited = context.get('visited', set())
+            context['visited'] = executor_visited.union(state_visited)
+
+            executor_instructions_count = context.get('instructions_count', 0)
+            context['instructions_count'] = executor_instructions_count + state_instructions_count 
+
+    def _fork_state_callback(self, state, expression, values, policy):
```
Comment (mossberg) [inline manticore/manticore.py]: imo, we should delete all these callbacks and their signals, if these stub callbacks are the only slots connected to them. that is, until we have a use case for them.

File: manticore/manticore.py [hunk 20]
```diff
@@ -505,7 +537,189 @@ def load_assertions(self, path):
...
+
+
+    def _emulate_instruction_callback(self, state, cpu, instruction):
+        logger.debug("About to emulate instruction")
+
+    def _did_execute_instruction_callback(self, state, cpu, instruction):
+        logger.debug("Did execute an instruction")
+
+    def _execute_instruction_callback(self, state, cpu, instruction):
+        logger.info("exe\n")
```
Comment (mossberg) [inline manticore/manticore.py]: i don't think this is a useful logging statement?

File: manticore/manticore.py [hunk 21]
```diff
@@ -505,7 +537,189 @@ def load_assertions(self, path):
...
+        address = state.cpu.PC
+        if not issymbolic(address):
+            state.context.setdefault('visited', set()).add(address)
+            count = state.context.get('instructions_count', 0)
+            state.context['instructions_count'] = count + 1
+
+
+    def _generate_testcase_callback(self, state, testcase_id, message = 'Testcase generated'):
+        #Fixme split this!
```
Comment (mossberg) [inline manticore/manticore.py]: we've been using a convention of `FIXME(name)`, can we do that here?

File: manticore/platforms/linux.py [hunk 22]
```diff
@@ -1397,7 +1394,7 @@ def sys_exit_group(self, error_code):
         #self.procs[procid] = None
         logger.debug("EXIT_GROUP PROC_%02d %s", procid, error_code)
         if len(self.running) == 0 :
-            raise ProcessExit(error_code)
+            raise Exception('Process exited correctly. Code: {}'.format(error_code))
```
Comment (mossberg) [inline manticore/platforms/linux.py]: why are we using a regular exception here instead of ProcessExit?

File: manticore/platforms/platform.py [hunk 23]
```diff
@@ -2,6 +2,29 @@
 from itertools import islice, imap
 import inspect
 
+class OSException(Exception):
```
Comment (mossberg) [inline manticore/platforms/platform.py]: this is duplicated in models/system.py?

File: manticore/platforms/windows.py [hunk 24]
```diff
@@ -48,9 +47,9 @@ def toStr(state, value):
             value = minmax[0]
     return '{:08x}'.format(value)
 
-class Windows(Platform):
+class Windows(object):
```
Comment (mossberg) [inline manticore/platforms/windows.py]: revert these changes too? windows should be a platform?

File: manticore/models/system.py (hunk 15 above)
Comment (feliam) [inline manticore/models/system.py]: This system.py is a desperate attempt to abstract the operating system models. Which think its now .. platform. So in the case that exception is actually used it should go somewhere in platform/. tbc..

File: manticore/platforms/platform.py (hunk 23 above)
Comment (feliam) [inline manticore/platforms/platform.py]: bye bye system.py.

File: manticore/core/cpu/abstractcpu.py [hunk 25]
```diff
@@ -387,7 +458,9 @@ def read_register(self, register):
         :return: register value
         :rtype: int or long or Expression
         '''
-        return self._regfile.read(register)
+        value = self._regfile.read(register)
+        self.will_read_register(register, value)
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Sort of inaccurate, `will_read_register` should execute before the actual `read`. (i.e. if `read` raises, the signal won't fire)

File: manticore/core/cpu/abstractcpu.py [hunk 26]
```diff
@@ -1,21 +1,76 @@
...
+        self.bytes=bytes
+
+class InstructionNotImplementedError(CpuException):
+    '''
+    Exception raised when you try to execute an instruction that is not yet
+    implemented in the emulator. Add it to the Cpu-specific implementation.
+    '''
+    pass
+
+class DivideError(CpuException):
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Rename to `DivideByZeroError`?

File: manticore/core/cpu/abstractcpu.py [hunk 27]
```diff
@@ -479,89 +554,98 @@ def read_bytes(self, where, size):
...
+            if issymbolic(c):
+                assert isinstance(c, BitVec) and  c.size == 8
+                if isinstance(c, Constant):
+                    c = chr(c.value)
+                else:
+                    logger.error('Concretize executable memory %r %r', c, text )
+                    raise ConcretizeMemory(self.memory, address = pc,
+                                            size = 8 * self.max_instr_width, 
+                                            policy = 'INSTRUCTION' )
+            assert isinstance(c, str)
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: What else can `c` be here?

File: manticore/core/cpu/abstractcpu.py [hunk 28]
```diff
@@ -479,89 +554,98 @@ def read_bytes(self, where, size):
...
-        instruction = next(self._md.disasm(code, pc))
 
-        #PC points to symbolic memory 
-        if instruction.size > len(text):
-            logger.info("Trying to execute instructions from invalid memory")
-            raise InvalidPCException(pc)
+        #decode the instructtion from code 
+        try:
+            instruction = next(self._md.disasm(code, pc))
+        except Exception as e:
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Should we catch only Capstone exceptions here?

File: manticore/core/executor.py [hunk 29]
```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
...
+        ''' Executor context is a shared memory object. All workers share this. 
+            It needs a lock. Its used like this:
+
+            with executor.context() as context:
+                vsited = context['visited']
+                visited.append(state.cpu.PC)
+                context['visited'] = visited
+        '''
+        with self._lock:
+            yield self._shared_context
```
Comment (yan) [inline manticore/core/executor.py]: What is the `_lock` protecting?

File: manticore/manticore.py (hunk 19 above)
Comment (feliam) [inline manticore/manticore.py]: Yes! also if not connected the signals wont propagate and stuff would be faster.
Not sure what would be the manticore API to hook on the different signals.

File: manticore/platforms/linux.py (hunk 22 above)
Comment (feliam) [inline manticore/platforms/linux.py]: Yeah :( .. ideas?
Import and raise state.TerminateState ?
Make a PlatformException for all the platforms and raise that?

File: manticore/core/cpu/abstractcpu.py (hunk 25 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: hmm will/did then ?
Only did_read_register ?
Or on_register_read?
I'm lost with the naming scheme here.

File: manticore/core/cpu/abstractcpu.py (hunk 27 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: Ideally it can't be anything else. 
But it could be an Expression if the concretizing code above failed. 
If that's double checked and we are certain I may just remove this.

File: manticore/core/executor.py (hunk 29 above)
Comment (feliam) [inline manticore/core/executor.py]: So _shared_context is shared between all running workers.
If we grant access to one without a lock and the callback at that worker do some non trivial read writes it may lead to a race. no? (there is already an example hook using shared_context at manticore)

File: manticore/core/executor.py [hunk 30]
```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
...
+                saved_states.append(self._getFilename(filename)) 
+        
+        #We didn't find any saved intermediate states in the workspace
+        if not saved_states:
+            return False
+
+        #search finalized testcases
+        saved_testcases = []
+        for filename in os.listdir(self.workspace):
+            if filename.startsswith('test_') and filename.endswith('.pkl'):
```
Comment (yan) [inline manticore/core/executor.py]: Typo? (`startsswith`)

File: manticore/core/executor.py [hunk 31]
```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
...
+                saved_testcases.append(self._getFilename(filename)) 
+
+        if saved_states or saved_testcases:
+            #We are trying to continue a paused analysis
+            if initial is not None:
+                raise Exception('Cant set initial state when continuing from previous workspace: {}'.format(workspace))
```
Comment (yan) [inline manticore/core/executor.py]: Raise a more specific exception?

File: manticore/core/executor.py [hunk 32]
```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
...
+            logger.debug("Waiting for available states")
+            self._lock.wait()
+            
+        state_id = random.choice(self._states)
+        del  self._states[self._states.index(state_id)]
+        return state_id
+
+    ###############################################################
+    # File Storage 
+    def backup(self, state):
```
Comment (yan) [inline manticore/core/executor.py]: What is state backup?

Comment (feliam) [inline manticore/core/executor.py]: Its when you save a State to the secondary storage(in our current implementation is a pickled file on the workspace folder).    backup/restore

File: manticore/core/cpu/abstractcpu.py (hunk 28 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: There is a CsError. But I couldn't make it throw it. 
I got 
StopIteration, ctypes.ArgumentError/TypeError


felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(CODE, "SAS"))
  File "/home/felipe/.local/lib64/python2.7/site-packages/capstone/__init__.py", line 849, in disasm
    res = _cs.cs_disasm(self.csh, code, len(code), offset, count, ctypes.byref(all_insn))
ctypes.ArgumentError: argument 4: <type 'exceptions.TypeError'>: wrong type
felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(2, 0x1000))
  File "/home/felipe/.local/lib64/python2.7/site-packages/capstone/__init__.py", line 849, in disasm
    res = _cs.cs_disasm(self.csh, code, len(code), offset, count, ctypes.byref(all_insn))
TypeError: object of type 'int' has no len()
felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(CODE, 0x1000,-1))
StopIteration

File: manticore/core/cpu/abstractcpu.py (hunk 18 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: This is how the callback for this signal will look like at manticore level. 
   def _execute_instruction_callback(self, state, platform, cpu, instruction):
        logger.info("exe\n")

Instruction will be available as an argument and as cpu.intruction both.

File: manticore/core/cpu/abstractcpu.py (hunk 4 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: All signals here are at least emitted. 
A use case for the read/write-register/memory could be: report the next time you "see" a tainted value. When a tainted value is read from memory or register.

When we emulate an instruction we lost read/write-register//memory so I assume that a user would want to know when those are not available. How? hooking on will_emulate/did_emulate

File: manticore/core/cpu/abstractcpu.py [hunk 33]
```diff
@@ -293,13 +347,15 @@ def syscall_number(self):
         '''
         raise NotImplementedError
 
+from ...utils.event import Signal
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Why not import during the imports at the beginning of file?

File: manticore/core/cpu/abstractcpu.py [hunk 34]
```diff
@@ -293,13 +347,15 @@ def syscall_number(self):
...
+
 ############################################################################
 # Abstract cpu encapsulating common cpu methods used by platforms and executor.
 class Cpu(object):
     '''
     Base class for all Cpu architectures. Functionality common to all
     architectures (and expected from users of a Cpu) should be here. Commonly
-    used by platforms and py:class:manticore.core.Executor
+    used by plaforms and py:class:manticore.core.Executor
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: typo

File: manticore/core/cpu/abstractcpu.py [hunk 35]
```diff
@@ -446,9 +534,13 @@ def read_int(self, where, size=None):
         if size is None:
             size = self.address_bit_size
         assert size in SANE_SIZES
+        self.will_read_memory(where, size)
+
         data = self.memory[where:where+size/8]
         total_size = 8 * len(data)
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: When will `len(data)*8` not equal `size`?

File: manticore/core/cpu/abstractcpu.py [hunk 36]
```diff
@@ -570,27 +676,40 @@ def fallback_to_emulate(*operands):
...
         if logger.level == logging.DEBUG :
             logger.debug(self.render_instruction())
             for l in self.render_registers():
                 register_logger.debug(l)
-
+        
         implementation(*instruction.operands)
         self._icount+=1
 
+        #broadcast event
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: i'd remove all the 'broadcast event' comments, as the naming convention is pretty clear and they add significant extra lines, imho. (also okay with leaving them in)

File: manticore/core/cpu/abstractcpu.py [hunk 37]
```diff
@@ -605,7 +724,15 @@ def render_instruction(self):
 
     def render_register(self, reg_name):
         result = ""
-        value = self.read_register(reg_name)
+
+        self.will_read_register.disable()
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: `render_register` is only called from `render_registers`. Move the disable/renable stuff there, otherwise we're disabling/reenabling all the signals a ton unnecessarily.

File: manticore/core/cpu/abstractcpu.py [hunk 38]
```diff
@@ -630,114 +761,13 @@ def __str__(self):
...
-
-
 #Instruction decorators
 def instruction(old_method):
     #This should decorate every instruction implementation
     @wraps(old_method)
     def new_method(cpu, *args, **kw_args):
         cpu.PC += cpu.instruction.size
-        return old_method(cpu,*args,**kw_args)
+        value = old_method(cpu,*args,**kw_args)
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Placeholder from old code? why not return inline?

File: manticore/core/cpu/abstractcpu.py [hunk 39]
```diff
@@ -479,89 +571,103 @@ def read_bytes(self, where, size):
...
-                    else:
-                        logger.error('Concretize executable memory %r %r', c, text )
-                        break
-                assert isinstance(c, str)
-                text += c
-        except MemoryException:
+                text += _read_concrete_byte(pc+i)
+        except MemoryException as mem_except:
+            #lets not raise this exception in case the instruction
+            #was fully read befor the page
```
Comment (yan) [inline manticore/core/cpu/abstractcpu.py]: Perhaps rewrite as:
```
try:
    # Read Instruction from memory
    for i in xrange(0, self.max_instr_width):
        try:
            text += _read_concrete_byte(pc+i)
        except MemoryException:
            # In case of reading into unmapped memory
            break
```
Otherwise, we throw+except for every byte until `max_instr_width`.

File: manticore/platforms/decree.py [hunk 40]
```diff
@@ -88,10 +91,10 @@ class Decree(object):
 
     def __init__(self, programs):
         '''
-        Builds a Decree OS platform
-        :param cpus: CPU for this platform.
-        :param mem: memory for this platform.
-        :todo: generalize for more CPUs.
+        Builds a Decree OS
+        :param cpus: CPU for this platform
+        :param mem: memory for this oplatform
```
Comment (yan) [inline manticore/platforms/decree.py]: typo?

File: manticore/manticore.py [hunk 41]
```diff
@@ -469,20 +497,24 @@ def apply_model_hooks(self, path):
...
         import core.cpu
         import importlib
-        import platforms
+        import models
 
         with open(path, 'r') as fnames:
             for line in fnames.readlines():
                 address, cc_name, name = line.strip().split(' ')
-                fmodel = platforms
+                cc = getattr(core.cpu.x86.ABI, cc_name)
```
Comment (yan) [inline manticore/manticore.py]: A lot of the code here seems reverted to much older code, including naming above (`platform` back to `model`)

File: manticore/utils/event.py [hunk 42]
```diff
@@ -1,13 +1,51 @@
...
+
+def forward_signals(dest, source):
+    ''' 
+        Replicate and forward all the signals from source to dest
+    '''
+    #Import all signals from state
+    for signal_name in dir(source):
+        if not signal_name.startswith('will') and \
+            not signal_name.startswith('did') and \
+            not signal_name.startswith('on'):
```
Comment (yan) [inline manticore/utils/event.py]: This might catch a lot of unintended fields. Isn't this unnecessary with the `isinstance` below?

File: manticore/manticore.py [hunk 43]
```diff
@@ -505,7 +537,193 @@ def load_assertions(self, path):
...
+        logger.debug("Decoding stuff instruction not available")
+
+
+    def _emulate_instruction_callback(self, state, platform, cpu, instruction):
+        logger.debug("About to emulate instruction")
+
+    def _did_execute_instruction_callback(self, state, platform, cpu, instruction):
+        logger.debug("Did execute an instruction")
+
+    def _execute_instruction_callback(self, state, platform, cpu, instruction):
```
Comment (mossberg) [inline manticore/manticore.py]: i'm not sure I like how each layer of signal forwarding adds on an extra parameter to the slot/callback. it seems like it could be redundant, for example, given that platform, cpu and instruction and all accessible from the state object?

i also think the way `forward_signal` setattrs makes the code a lot harder to read since it is totally unclear how that signal attribute exists on the class otherwise.

Comment (mossberg) [inline manticore/manticore.py]: would it be possible to have some sort of global event system rather than a hierarchical chain?

File: manticore/core/cpu/abstractcpu.py (hunk 35 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: 100 != len("ABC"[0:100])

File: manticore/core/cpu/abstractcpu.py (hunk 37 above)
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: That's it I'll remove the cli/sti from there. Lets document it and force the user to do it explicitly for now.  I'm not really fixing anything fully with disabling it just here.

File: manticore/manticore.py (hunk 43 above)
Comment (feliam) [inline manticore/manticore.py]: Forwarding Objects are saved internally at Signal anyway. We could omit them in them in the emit()
We can add an extra keyword argument to Signal.when() to control if the forwarding object has to interpose itself in the callback arguments or not.  
Idea is to have only state + specific arguments at the callback, right?

---- 
For the global event system. You mean that every object holds a reference to .events of type Events()
class Events():
   register_new_signal(str, desc)
   register_callback(str, obj, method)
   list_signals()
   disable()
   enable()

?

File: manticore/core/cpu/abstractcpu.py [hunk 44]
```diff
@@ -144,16 +146,7 @@ def size(self):
...
-    def type(self):
-        ''' This property encapsulate the operand type. 
-            It may be one of the following:
-                register
-                memory
-                immediate
-        '''
-        raise NotImplementedError
-        
+    @abstractmethod
```
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: Better not use ABC. So we do not have the overhead of checking in each invocation.

File: manticore/core/cpu/abstractcpu.py [hunk 45]
```diff
@@ -144,16 +146,7 @@ def size(self):
     def reg(self):
         return self._reg_name(self.op.reg)
 
-    @property
-    def type(self):
```
Comment (feliam) [inline manticore/core/cpu/abstractcpu.py]: Better to document in the abstract class so implementations overload this?
